README.md
# Tempus [![PyPI](https://img.shields.io/pypi/v/tempus.svg)](https://pypi.python.org/pypi/tempus) [![Travis](https://img.shields.io/travis/chunkhang/tempus.svg)](https://travis-ci.org/chunkhang/tempus)

<img src="https://images.unsplash.com/photo-1501139083538-0139583c060f?auto=format&fit=crop&w=1950&q=60&ixid=dW5zcGxhc2guY29tOzs7Ozs%3D" alt="Hourglass" width=400/><br/>
***TEMPUS EST DE ESSENTIA*** <br/>
*Time is of the essence*

## INSTALLATION

```
$ pip install tempus
```
**Python Versions** <br/>
3.3 | 3.4 | 3.5 | 3.6

## USAGE

### Timer
**tempus timer \<duration\>** <br/>
tempus timer 1m | tempus timer 2m15s | tempus timer 30s | tempus timer 10000s <br/><br/>
![Tempus Timer](https://user-images.githubusercontent.com/12708862/32622525-188160a2-c5be-11e7-937d-6eb5c8209a22.gif)

### Stopwatch

Coming soon

### Metronome

Coming soon

## CAVEATS

### Font
The progress bar may appear distorted or unsmooth when filling up due to the terminal's font not displaying the unicode characters as intended. Hence, you may need to play around with your font to get the best experience.
```
$ tempus test
Ensure that the characters below are:
* Equal in height
* Forming a smooth gradient across
* Aesthetically pleasing

|█▉▊▋▌▍▎▏ |
|⣿⣷⣧⣇⡇⡆⡄⡀|

https://github.com/chunkhang/tempus/#font
```
**Recommended Fonts** <br/>
Menlo | SF Mono <br/>

### Terminal
The progress bar needs a certain amount of space to ensure proper printing. Therefore, you may get a message if your terminal size is too small. In that case, just adjust your terminal's window before executing the command again.
```
$ tempus timer 1m
The terminal is too small: 52 x 25
It needs to be at least  : 80 x 10
```
**Minimum Size** <br/>
80 x 10

## RELEASES

### 1.1.0
* Shiny new progress bar for timer

### 1.0.0
* Initial release

## UNLICENSE

```
This is free and unencumbered software released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <http://unlicense.org>
```

## CONTACT

Feel free to report bugs or suggest features <br/>
**[Marcus Mu](http://marcusmu.me)** | chunkhang@gmail.com


CHANGELOG
1.0.0
-----
- Initial release

1.1.0
-----
- Added progress bar for timer
- Added test command
- Added utils module with test
- Updated readme significantly


test/test_timer.py
import sys; sys.path.append('../tempus')
from tempus.commands import timer

def test_valid_syntax():
   assert timer._syntax_valid('1m2s') == True
   assert timer._syntax_valid('10m59s') == True
   assert timer._syntax_valid('1m2s') == True
   assert timer._syntax_valid('0m2s') == True
   assert timer._syntax_valid('1m0s') == True
   assert timer._syntax_valid('0m0s') == True
   assert timer._syntax_valid('1000m2s') == True
   for i in range(60):
      assert timer._syntax_valid('{}s'.format(i)) == True

def test_invalid_syntax():
   assert timer._syntax_valid('10') == False
   assert timer._syntax_valid('1M') == False
   assert timer._syntax_valid('2S') == False
   assert timer._syntax_valid('1m10s10') == False
   assert timer._syntax_valid('-1m-2s') == False
   assert timer._syntax_valid('-1m') == False
   assert timer._syntax_valid('-2s') == False

def test_to_duration():
   assert timer._to_duration('10m59s') == (10, 59)
   assert timer._to_duration('0m0s') == (0, 0)
   assert timer._to_duration('1m') == (1, None)
   assert timer._to_duration('1s') == (None, 1)
   assert timer._to_duration('') == (None, None)

def test_valid_duration():
   assert timer._values_valid((1, 1)) == True
   assert timer._values_valid((1, 0)) == True
   assert timer._values_valid((0, 1)) == True
   assert timer._values_valid((1000, 59)) == True
   for i in range(60):
      assert timer._values_valid((40, i)) == True

def test_invalid_duration():
   assert timer._values_valid((0, 0)) == False
   assert timer._values_valid((0, 60)) == False
   assert timer._values_valid((0, 61)) == False
   assert timer._values_valid((0, 61)) == False
   assert timer._values_valid((0, None)) == False
   assert timer._values_valid((None, 0)) == False
   assert timer._values_valid((None, None)) == False

def test_time_string():
   assert timer.Bar.Timer(0, None)._to_string(0) == '0:00'
   assert timer.Bar.Timer(20, None)._to_string(20) == '0:20'
   assert timer.Bar.Timer(720, None)._to_string(720) == '12:00'
   assert timer.Bar.Timer(721, None)._to_string(721) == '12:01'

test/test_utils.py
import sys; sys.path.append('../tempus')
from tempus import utils

def test_truncate():
   assert utils.truncate(1.67592, 0) == 1
   assert utils.truncate(1.67592, 1) == 1.6
   assert utils.truncate(1.67592, 2) == 1.67
   assert utils.truncate(1.67592, 3) == 1.675
   assert utils.truncate(1.67592, 4) == 1.6759
   assert utils.truncate(1.67592, 5) == 1.67592
   assert utils.truncate(1.67592, 6) == 1.67592

tempus/commands/test.py
from click import command, argument

from . import timer
from .. import constant

@command('test', short_help='test font suitability')
def execute(): 
   print('Ensure that the characters below are:')
   print('* Equal in height')
   print('* Forming a smooth gradient across')
   print('* Aesthetically pleasing')
   print()
   for v in constant.FRAME_CHAR_STAGES.values():
      print('{}{}{}'.format(
         constant.BORDER_CHAR,
         ''.join(reversed(v)),
         constant.BORDER_CHAR))
   print()
   print('https://github.com/chunkhang/tempus/#font')

tempus/utils.py
# utils.py
# Utility functions

import math
import os
import shlex
import struct
import platform
import subprocess

from . import constant

def truncate(float_number, decimal_point):
   '''
   e.g. _truncate(1.67592, 3) -> 1.675
   '''   
   magic_number = 10 ** decimal_point
   return math.floor(float_number*magic_number) / magic_number

def terminal_size(_default=constant.TERMINAL_DEFAULT_SIZE):
   '''
   -> (int, int): (width, height)
   Code obtained and modified from: https://gist.github.com/jtriley/1108174
   '''
   def _get_terminal_size_windows():
      try:
         from ctypes import windll, create_string_buffer
         # stdin handle is -10
         # stdout handle is -11
         # stderr handle is -12
         h = windll.kernel32.GetStdHandle(-12)
         csbi = create_string_buffer(22)
         res = windll.kernel32.GetConsoleScreenBufferInfo(h, csbi)
         if res:
            (bufx, bufy, curx, cury, wattr,
               left, top, right, bottom,
               maxx, maxy) = struct.unpack("hhhhHhhhhhh", csbi.raw)
            sizex = right - left + 1
            sizey = bottom - top + 1
            return sizex, sizey
      except:
         pass
   def _get_terminal_size_tput():
      try:
         cols = int(subprocess.check_output(shlex.split('tput cols')))
         rows = int(subprocess.check_output(shlex.split('tput lines')))
         return (cols, rows)
      except:
         pass
   def _get_terminal_size_linux():
      def ioctl_GWINSZ(fd):
         try:
            import fcntl
            import termios
            cr = struct.unpack('hh', 
               fcntl.ioctl(fd, termios.TIOCGWINSZ, '1234'))
            return cr
         except:
            pass
      cr = ioctl_GWINSZ(0) or ioctl_GWINSZ(1) or ioctl_GWINSZ(2)
      if not cr:
         try:
            fd = os.open(os.ctermid(), os.O_RDONLY)
            cr = ioctl_GWINSZ(fd)
            os.close(fd)
         except:
            pass
      if not cr:
         try:
            cr = (os.environ['LINES'], os.environ['COLUMNS'])
         except:
            return None
      return int(cr[1]), int(cr[0])
   current_os = platform.system()
   tuple_xy = None
   if current_os == 'Windows':
      tuple_xy = _get_terminal_size_windows()
      if tuple_xy is None:
         tuple_xy = _get_terminal_size_tput()
   if current_os in ['Linux', 'Darwin'] or current_os.startswith('CYGWIN'):
      tuple_xy = _get_terminal_size_linux()
   if tuple_xy is None:
      tuple_xy = _default
   return tuple_xy


tempus/commands/timer.py
# -*- coding: utf-8 -*-

from click import command, argument
import cursor
import time
import re
import sys
import threading
import itertools

from .. import constant
from ..utils import truncate, terminal_size

@command('timer', short_help='use timer tool')
@argument('duration')
def execute(duration):

   # Check if terminal is at least minimum size
   size = terminal_size()
   minimum = constant.TERMINAL_MINIMUM_SIZE
   if not all(map(lambda x, y: x >= y, size, minimum)):
      print('The terminal is too small: {} x {}'.format(size[0], size[1]))
      print('It needs to be at least  : {} x {}'.format(minimum[0], minimum[1]))
      sys.exit()

   # Check duration syntax
   if _syntax_valid(duration):
      # Check duration values
      duration = _to_duration(duration)
      if not _values_valid(duration):
         print('Invalid duration.')
         print('Examples: 2m59s, 40m0s, 0m10s')
         sys.exit()
   else:
      print('Invalid syntax.')
      print('Examples: 10s, 5m, 20m30s')
      sys.exit()      

   # Start timer bar
   input('<Enter> to start')
   print()
   Bar(_to_seconds(duration)).start()
   print()

   # Ring bell
   _ring_bell('<Ctrl-C> to stop')
   sys.exit()
   

def _syntax_valid(ds, r=constant.DURATION_REGEX):
   '''
   -> boolean: True - valid syntax
   -> boolean: False - invalid syntax
   '''
   m = re.compile(r).match(ds)
   if m:
      return True
   else:
      return False

def _to_duration(ds, r=constant.DURATION_REGEX):
   '''
   -> tuple: (int/None, int/None) - (minutes, seconds)
   '''
   g = re.compile(r).match(ds).groups()
   m = int(g[1]) if g[1] is not None else None
   s = int(g[3]) if g[3] is not None else None
   return (m, s)

def _values_valid(d):
   '''
   -> boolean: True - valid duration
   -> boolean: False - invalid duration
   '''
   m, s = d
   # No minutes and seconds
   if m is None and s is None:
      return False
   # Minutes and seconds
   if m is not None and s is not None:
      if (m == 0 and s == 0) or \
         (s >= 60):
         return False
   # Minutes only
   elif s is None:
      if m == 0:
         return False
   # Seconds only
   elif m is None:
      if s == 0:
         return False
   return True

def _to_seconds(d):
   '''
   Convert duration string to seconds
   -> int: total seconds 
   '''
   m, s = d
   m = 0 if m is None else m
   s = 0 if s is None else s
   return m*60 + s

class Bar(object):
   '''
   Class for progress bar
   '''

   class Updater(threading.Thread):
      '''
      Class to continuously update bar until done
      '''
      def __init__(self, bar, timer):
         threading.Thread.__init__(self)
         self._bar = bar
         self._timer = timer
         self._stop_now = False
      def run(self):
         last = False
         while True:
            if self._stop_now:
               break
            if self._bar._get_done():
               last = True
            bar = self._bar._get_bar()
            time = self._timer.get_current_time() if not last else \
               self._timer.get_zero_time()
            sys.stdout.write('{}{}{} {} {}'.format(
               constant.BORDER_CHAR, 
               bar,
               constant.BORDER_CHAR,
               time,
               constant.BORDER_CHAR))
            sys.stdout.write('\r')
            sys.stdout.flush()
            if last:
               sys.stdout.write('\n')
               sys.stdout.flush()
               break
      def stop(self):
         self._stop_now = True

   class Timer(threading.Thread):
      '''
      Class to update time remaining
      '''
      def __init__(self, starting_seconds, bar):
         threading.Thread.__init__(self)
         self._initial_length = 0
         self._initial_length = len(self._to_string(starting_seconds))
         self._remaining_seconds = starting_seconds
         self._bar = bar
         self._stop_now = False
      # Convert seconds to string
      def _to_string(self, seconds):
         m, s = divmod(seconds, 60)
         time_string = '{}:{:02d}'.format(m, s)
         if self._initial_length == 0:
            return time_string
         else:
            return time_string.rjust(self._initial_length)
      # Start counting down
      def run(self):
         while self._remaining_seconds > 0:
            if self._stop_now:
               break
            time.sleep(1)
            self._remaining_seconds -= 1
         self._bar.notify()
      def stop(self):
         self._stop_now = True
      def get_current_time(self):
         return self._to_string(self._remaining_seconds)
      def get_zero_time(self):
         return self._to_string(0)

   def __init__(self, duration, frame_char_type='block', length=80):
      threading.Thread.__init__(self)
      # Duration
      self._timer = Bar.Timer(duration, self)
      self._total_seconds = duration
      self._time_up = False
      self._done = False
      # Frames
      self._updater = Bar.Updater(self, self._timer)
      self._total_frames = length - 5 - len(self._timer.get_current_time())
      self._current_frame = 1
      self._frame_stack = ''
      self._bar = constant.REMAINING_CHAR*self._total_frames
      # Stages
      frame_stages = constant.FRAME_CHAR_STAGES[frame_char_type]
      self._stage_in_cycle = itertools.cycle(frame_stages)
      self._last_stage = frame_stages[-1]
      # Gaps (In seconds)
      stages_per_frame = len(frame_stages)
      total_gaps = self._total_frames*stages_per_frame - 1
      self._normal_gap = truncate(self._total_seconds/total_gaps, 5)
      self._first_gap = \
         self._total_seconds - total_gaps*self._normal_gap + self._normal_gap    

   # Return complete bar for updater
   def _get_bar(self):
      return self._bar

   # Return done status
   def _get_done(self):
      return self._done

   # Start progress bar
   def start(self):
      with cursor.HiddenCursor():
         first_stage_now = True
         remaining_frames = self._total_frames - self._current_frame
         try:
            # Start timer and updater threads
            self._timer.start()
            self._updater.start()
            while self._current_frame <= self._total_frames:
               # Add frames stage by stage smoothly, if there is still time
               if not self._time_up:
                  # Sleep between stages
                  if first_stage_now:
                     time.sleep(self._first_gap)
                     first_stage_now = False
                  else:
                     time.sleep(self._normal_gap)
                  # Update bar
                  current_stage = next(self._stage_in_cycle)
                  self._bar = self._frame_stack + current_stage + \
                     constant.REMAINING_CHAR*remaining_frames
                  # Update frame stack when the current frame is complete
                  # Move on to next frame
                  if current_stage == self._last_stage:
                     self._frame_stack += self._last_stage
                     self._current_frame += 1
                     remaining_frames -= 1
               # Fill in remaining frames immediately, if time is up
               else:
                  self._bar = self._last_stage*self._total_frames
                  break
            self._done = True
         except KeyboardInterrupt:
            # Stop threads
            self._timer.stop()
            self._updater.stop()
            sys.exit()
         # Wait for threads to finish
         self._timer.join()
         self._updater.join()

   # Receive notification that time is up
   def notify(self):
      self._time_up = True

def _ring_bell(message, frequency=0.5):
   def _beep():
      print('\a', end='', flush=True)
   first_ring = True
   while True:
      try:
         if first_ring:
            print(message, end='')
            first_ring = False
         _beep()
         time.sleep(frequency)
      except KeyboardInterrupt:
         print()
         break